src/background/EventManager.ts

```diff
@@ -46,34 +28,41 @@ export class EventManager {
       types: ["main_frame"],
     };
     const extraInfoSpec: WebRequest.OnHeadersReceivedOptions[] = ["blocking"];
-    this.webRequestHeadersEmitter.addListener(
-      this.receiveWebRequest.bind(this),
+
+    this.webRequest.onBeforeRequest.addListener(
+      this.resetTabData.bind(this),
+      filter,
+      []
+    );
+    this.webRequest.onHeadersReceived.addListener(
+      this.receiveWebRequestHeaders.bind(this),
       filter,
       extraInfoSpec
     );
-    this.webRequestErrorEmitter.addListener(
+    this.webNavigation.onErrorOccurred.addListener(
       this.receiveWebRequestError.bind(this)
     );
-    this.messageEmitter.addListener(this.receiveMessage.bind(this));
-    this.tabActivatedEmitter.addListener(this.changeBrowserAction.bind(this));
-    this.webRequestBeforeEmitter.addListener(
-      this.resetTabData.bind(this),
-      filter,
-      []
-    );
+    this.runtime.onMessage.addListener(this.receiveMessage.bind(this));
+    this.tabs.onActivated.addListener(this.changeBrowserAction.bind(this));
   }
 
   public resetTabData(requestDetails: { tabId: number }): void {
     this.app.resetTabData(requestDetails.tabId);
   }
 
-  public receiveWebRequest(
+  public async receiveWebRequestHeaders(
     requestDetails: WebRequest.OnHeadersReceivedDetailsType
-  ): void {
-    // using await is not possible here, since making receiveWebRequest async is not allowed
-    this.app.fetchCertificate(requestDetails).then(() => {
-      this.changeBrowserAction(requestDetails);
-    });
+  ): Promise<WebRequest.BlockingResponse> {
+    const hasQualityDecreased = await this.app.fetchCertificate(requestDetails);

```

Comment:
Find another solution for getting this information please. Maybe await the fetch and the call a Method `hasQualityDecreased` on the app with the request details?

It is not clear that `fetchCertificate` returns a boolean regarding certificate quality.

Reply:
Has to be documented well since the method name does not make it clear that the return value will be a boolean?